### PR TITLE
[SPARK-31473][SQL] AQE should set active session during execution

### PR DESCRIPTION
**Pull request by maryannxue**
### What changes were proposed in this pull request?

AQE creates new SparkPlan nodes during execution. This PR makes sure that the active session is set correctly during this process and AQE execution is not disrupted by external session change.

### Why are the changes needed?

To prevent potential errors. If not changed, the physical plans generated by AQE would have the wrong SparkSession or even null SparkSession, which could lead to NPE.

### Does this PR introduce any user-facing change?

No.

### How was this patch tested?

Added UT.

### CODE REVIEW

**Comment by maryannxue** [comment]
cc @cloud-fan @JkSelf @Ngone51

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala [hunk 1]
```diff
@@ -139,7 +139,9 @@ case class AdaptiveSparkPlanExec(
   }
 
   private def getFinalPhysicalPlan(): SparkPlan = lock.synchronized {
-    if (!isFinalPlan) {
+    if (isFinalPlan) return currentPhysicalPlan
+
+    context.session.withActive {
```
Comment (hvanhovell) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala:144]: Maybe add a comment why this is needed. Something in the spirit of `Adaptive execution can replan a query outside of withActive scope functions (e.g. plan.queryExecution.rdd and that is why we need to set it here.`

File: sql/core/src/test/scala/org/apache/spark/sql/execution/adaptive/AdaptiveQueryExecSuite.scala [hunk 2]
```diff
@@ -907,4 +908,13 @@ class AdaptiveQueryExecSuite
       assert(plan.asInstanceOf[DataWritingCommandExec].child.isInstanceOf[AdaptiveSparkPlanExec])
     }
   }
+
+  test("AQE should set active session during execution") {
+    withSQLConf(SQLConf.ADAPTIVE_EXECUTION_ENABLED.key -> "true") {
+      val df = spark.range(10).select(sum('id))
+      assert(df.queryExecution.executedPlan.isInstanceOf[AdaptiveSparkPlanExec])
+      SparkSession.setActiveSession(null)
```
Comment (hvanhovell) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/adaptive/AdaptiveQueryExecSuite.scala:916]: Undo this after the test? Or are we setting it in before()?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/adaptive/AdaptiveQueryExecSuite.scala [hunk 3]
```diff
@@ -907,4 +908,13 @@ class AdaptiveQueryExecSuite
...
+      checkAnswer(df, Seq(Row(45)))
```
Comment (hvanhovell) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/adaptive/AdaptiveQueryExecSuite.scala:917]: This throws an NPE right without setting the session right?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/adaptive/AdaptiveQueryExecSuite.scala (hunk 2 above)
Comment (maryannxue) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/adaptive/AdaptiveQueryExecSuite.scala:916]: I believe it is somehow properly (secretly) set for each test. But you are right, I should undo it anyway.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/adaptive/AdaptiveQueryExecSuite.scala (hunk 3 above)
Comment (maryannxue) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/adaptive/AdaptiveQueryExecSuite.scala:917]: Yep.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Thank you, @maryannxue and @hvanhovell .
Merged to master/3.0.

**Pull request closed without merging**